### PR TITLE
fix: eliminate arrow z-order regression and pan/zoom jank

### LINKED ISSUE #92: fix(frontend): 화살표 첫 렌더 누락 + 캔버스 스크롤/이동 버벅임
## 목표:
자동 마인드맵에서 발견된 두 가지 렌더링 버그를 수정해 사용자 경험을 회복한다.
1. 화살표가 분석 결과 첫 렌더 시 그려지지 않고, 노드를 한 번 움직여야 비로소 그려지는 문제
2. 캔버스 스크롤·노드 이동 시 버벅임

## 작업 범위:
- [ ] **화살표 첫 렌더 누락 진단·수정** — Excalidraw arrow binding이 첫 `updateScene`에서 geometry를 적용하지 않는 것으로 추정. `startBinding`/`endBinding`이 첫 변환 시점에 갖춰지는지 확인
- [ ] **스크롤/이동 버벅임 진단·수정** — `convertToExcalidrawElements({ regenerateIds: false })` 결과 element가 `version`/`versionNonce`/`updated` 등 메타필드를 누락해 Excalidraw 내부 hit-test/cache가 매 프레임 재계산되는 것으로 의심
- [ ] **진단 단계 비교** — `api.getSceneElements()[0]`와 사용자가 직접 그린 `rectangle`의 필드 비교, Excalidraw 내부에서 어떤 필드가 hit-test cache 키로 사용되는지 확인
- [ ] **sticky connector에도 동일 수정 적용** — `frontend/src/sticky/sticky.ts`. 단 PR #91 머지 후 진행
- [ ] **단위 테스트 추가** — arrow의 binding이 첫 변환 시점에 `startBinding`/`endBinding`을 갖춰 반환되는지, 메타필드(`version` 등)가 채워지는지 회귀 테스트

## 완료 기준:
- 분석 실행 직후 화살표가 즉시 캔버스에 표시된다 (노드 이동 없이)
- 캔버스에서 휠 스크롤·노드 드래그 시 시각적 끊김(1회 frame skip 이상) 없음
- sticky connector 화살표도 동일하게 동작
- 단위 테스트 통과

## 참고:
- 사용자 보고: PR #81 머지 후 webview 검증 시 두 현상 발견
- PR #69(자동 마인드맵 렌더)와 PR #81(sticky)에서 도입된 `convertToExcalidrawElements` 사용 패턴이 공통 의심 지점
- PR #91(sticky round-trip)이 `sticky.ts`/`App.tsx`를 광범위하게 변경 중 → 머지 충돌 회피 위해 #91 머지 후 진행

## 브랜치:
`fix/arrow-render-and-scroll-perf`

## 의존:
- PR #91 머지 (`sticky.ts`/`App.tsx` 충돌 회피)

### PR DESCRIPTION
**Pull request by jinsh1210**
## Summary
- 화살표가 노드 블록 위로 그려지는 z-order 회귀 수정: 엣지/노드 스켈레톤을 분리해 edges-first로 연결, 화살표가 노드 뒤에 렌더링되도록 변경
- 기존 저장된 `.codetrace` 파일 로드 시에도 z-order 보장: `toExcalidrawInitialData`에 `sortEdgesBehindNodes` 적용
- `routeAroundNodes` 좌표 버그 수정: 수직 세그먼트 우회 계산 시 Y 오프셋이 `midX`에 잘못 할당되던 문제 수정
- pan/zoom 시 단일 프레임 히컵 제거: `onChange` 동기 직렬화를 200ms 디바운스로 교체하고 pan/zoom skip 프레임에서 pending 타이머를 defer 처리

Closes #92

## Verification
- npm run build --workspace=frontend
- npm test --workspace=frontend -- --runInBand
- npm exec --workspace=frontend -- tsc --noEmit

### CODE REVIEW

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -560,21 +599,32 @@ export default function App() {
...
+        saveDebounceTimerRef.current = null;
+        pendingSaveCallbackRef.current = noop;
+        const document = createCanvasDocumentFromScene({
+          elements: capturedElements as unknown as ExcalidrawElementStub[],
+          appState: capturedAppState as unknown as Record<string, unknown>,
+          files: capturedFiles as unknown as Record<string, unknown>,
+        });
+        const content = serializeCanvasDocument(document);
+        if (content === latestContentRef.current) return;
+        latestContentRef.current = content;
```
Comment (babytazo) [inline frontend/src/App.tsx:669]: [P1] `latestContentRef`가 이제 이 debounced `doSave` 안에서만 최신 content로 갱신됩니다. 그래서 사용자가 element 변경 직후 200ms 안에 Ctrl+S를 누르면 `handleSaveShortcut`은 이전 `latestContentRef.current`를 `saveFile`로 보내고, 이후 timer는 최신 content를 일반 `save`로만 반영해서 디스크에는 이전 내용이 저장될 수 있습니다. pending save가 있으면 Ctrl+S에서 먼저 flush한 뒤 `saveFile`을 호출하거나, 직렬화 결과/dirty snapshot을 별도 ref로 즉시 갱신하는 식으로 수동 저장 경로와 debounce 경로를 맞춰야 합니다.

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -492,10 +497,42 @@ export default function App() {
...
+          appState.zoom.value !== prevApp.zoom.value) &&
+        appState.editingTextElement === prevApp.editingTextElement
+      ) {
+        // Defer any pending save so JSON.stringify doesn't fire mid-pan.
+        if (saveDebounceTimerRef.current !== null) {
+          clearTimeout(saveDebounceTimerRef.current);
+          saveDebounceTimerRef.current = setTimeout(pendingSaveCallbackRef.current, 200);
+        }
+        previousAppStateRef.current = appState;
+        return;
```
Comment (babytazo) [inline frontend/src/App.tsx:573]: [P1] pan/zoom fast path가 `elements.length`와 selection만 보고 return하기 때문에, auto-scroll 중 drag처럼 viewport 변화와 같은 길이의 element 변경이 같은 `onChange` 프레임에 들어오면 normalization, deletion detection, save scheduling을 전부 건너뜁니다. element id/version/updated timestamp 등 실제 scene 변경 여부를 비교하지 않으면 사용자의 이동/수정이 저장되지 않거나 `suppressDeletionDetectionRef` 상태가 다음 변경에 잘못 영향을 줄 수 있습니다.

File: frontend/src/graph/converter.ts [hunk 3]
```diff
@@ -122,15 +127,153 @@ export function convertGraphToElements(
...
+    midX = Math.abs(leftX - midX) <= Math.abs(rightX - midX) ? leftX : rightX;
+    break;
+  }
+
+  return [
+    [startRight, startCy],
+    [midX, startCy],
+    [midX, endCy],
+    [endLeft, endCy],
+  ];
```
Comment (babytazo) [inline frontend/src/graph/converter.ts:348]: [P2] 우회 로직이 `midX`의 수직 segment만 검사한 뒤 수평 segment는 검증하지 않습니다. 예를 들어 같은 y-row에서 A → B 사이에 중간 노드가 있고 `midX`가 그 노드를 잘라서 left/right로 이동하면, 반환되는 `[startRight,startCy] -> [midX,startCy]` 또는 `[midX,endCy] -> [endLeft,endCy]` 수평 leg가 그대로 중간 노드 내부를 지나갑니다. #92의 “노드 body를 통과하지 않는 라우팅” 보장을 위해 전체 polyline segment를 obstacle against로 검증하거나 dagre waypoints를 실제 렌더 경로로 사용해야 합니다.

File: frontend/src/storage/canvasStorage.ts [hunk 4]
```diff
@@ -58,6 +58,28 @@ export function toExcalidrawInitialData(document: CanvasDocument): ExcalidrawIni
...
+    const kind = (el.customData as { kind?: string } | undefined)?.kind;
+    if (kind === 'graphEdge') {
+      autoEdges.push(el);
+    } else if (kind === 'graphNode') {
+      autoNodes.push(el);
+    } else {
+      rest.push(el);
+    }
+  }
+  return [...autoEdges, ...autoNodes, ...rest];
```
Comment (babytazo) [inline frontend/src/storage/canvasStorage.ts]: [P2] 이 정렬은 저장된 scene 전체를 `graphEdge[] + graphNode[] + rest[]`로 재배치해서, `reviewSticky`나 사용자 화살표/텍스트처럼 graph가 아닌 element도 로드할 때마다 모든 graph node 위로 올라갑니다. 주석은 “without touching user elements”라고 되어 있지만 실제로는 persisted z-order를 크게 바꾸므로, auto graph edge가 관련 graph node보다 앞에 있을 때만 그 edge를 node 뒤로 내리는 식으로 non-graph element의 상대 순서는 보존하는 편이 안전합니다.

**Review by babytazo — CHANGES_REQUESTED** [review]
1차 리뷰입니다.

최신 head `1088908` 기준으로 다시 검토했습니다. base가 `develop`으로 정리되고, 이전에 보였던 `CanvasDocument` v2/cards 회귀는 제거된 것을 확인했습니다. CI도 최신 run 기준으로 unit/integration/extension-e2e 모두 통과했습니다.

다만 이번 PR의 핵심인 저장 디바운스와 화살표 라우팅 쪽에 실제 데이터 손실/회귀 가능성이 남아 있어 request changes로 남깁니다. 특히 `latestContentRef`가 debounce callback 이후에만 최신화되면서 Ctrl+S/외부 업데이트와 경합할 수 있고, pan/zoom fast path가 element count만 보고 같은 길이의 실제 scene 변경을 건너뛸 수 있습니다. 라우팅도 수직 segment만 검사해서 수평 leg가 중간 노드를 통과하는 케이스가 남습니다.

추가로 `frontend/src/App.tsx:325`의 기존 sticky merge short-circuit도 확인이 필요합니다. 새 connector geometry 보정은 `createStickyForAnchor`를 통해 재생성되는 경우에만 적용되는데, 저장된 sticky group이 connector를 이미 가지고 있으면 그대로 유지되어 예전 placeholder geometry가 수리되지 않을 수 있습니다.

아래 코멘트 반영 후 다시 확인하겠습니다.

**Comment by jinsh1210** [comment]
리뷰 반영했습니다. 최신 head `82941a6` 기준입니다.

---

**[P1] Ctrl+S vs debounce 경합 (`App.tsx:302`)**

`handleSaveShortcut`에서 `saveDocumentFile` 호출 전 `pendingSaveCallbackRef.current()`를 동기 flush 후 타이머를 clear합니다. Ctrl+S는 사용자 명시 동작이므로 동기 직렬화가 적절하고, 이후 `latestContentRef.current`가 최신 content를 가리키므로 `saveDocumentFile`에 전달되는 값이 항상 최신입니다.

**[P1] pan/zoom fast path element version 미검증 (`App.tsx:512`)**

`elements.length` 비교에 더해 element `version` 합산을 fingerprint로 비교합니다. 같은 길이의 scene 변화(auto-scroll 중 drag 등)가 같은 onChange 프레임에 들어와도 version sum이 달라지므로 normalization / deletion detection / save scheduling을 건너뛰지 않습니다. id별 version map 비교는 O(n) + Map overhead가 있고 sum도 충돌 가능성이 충분히 낮아 sum으로 결정했습니다.

**[P2] `sortEdgesBehindNodes` non-graph 상대 순서 깨짐 (`canvasStorage.ts:66`)**

전체를 세 그룹으로 재배치하는 방식에서, graph element가 점유하던 슬롯만 재배치하는 방식으로 교체했습니다. non-graph element(`reviewSticky`, `userArrow`, `userText` 등)는 원래 인덱스를 그대로 유지합니다.

**[P2] 수평 leg 노드 통과 (`converter.ts:222`)**

`routeAroundNodes`를 전체 3개 segment(상단 수평 + 수직 + 하단 수평) 모두 obstacle 검사하도록 교체했습니다. `hSegmentClipsNode` / `vSegmentClipsNode` 헬퍼를 추가하고, clipping node가 남아있는 동안 iterative하게 midX를 이동합니다. 회귀 테스트: 중간 노드가 default midX 경로 위에 있을 때 수평 leg도 통과하지 않음을 검증합니다.

**[리뷰 본문] sticky merge short-circuit placeholder connector (`App.tsx:325`)**

`isPlaceholderConnector` 헬퍼로 x=0, y=0, height=0, 2-point horizontal line인 connector를 감지합니다. 해당 connector를 가진 기존 group은 short-circuit을 통과시켜 `createStickyForAnchor`로 재생성하므로 예전 placeholder geometry가 수리됩니다.

---

추가로 `vite-env.d.ts`가 누락되어 IDE에서 `import './App.css'`에 타입 에러가 표시되던 문제도 함께 수정했습니다.

**Review by babytazo — CHANGES_REQUESTED** [review]
2차 리뷰입니다.

최신 head `82941a6` 기준으로 확인했습니다. 1차에서 지적했던 `Ctrl+S` stale save는 pending debounce를 flush하도록 고쳐졌고, `sortEdgesBehindNodes`도 graph slot만 재배치하는 방식이라 sticky/user element를 통째로 뒤로 미는 회귀는 해소된 것으로 봤습니다. sticky placeholder connector 재생성도 의도한 케이스는 처리됩니다.

다만 아직 approve하기에는 두 가지가 남았습니다.

1. `frontend/src/App.tsx:668` - 여기서 예약한 `doSave`는 `applyDocumentContent`로 새 content가 들어와도 취소되거나 무효화되지 않습니다. local 편집 직후 200ms 안에 extension/document update가 들어오면 `latestContentRef`는 새 content로 갱신되지만, 기존 타이머가 나중에 실행되면서 캡처해 둔 이전 `capturedElements/appState/files`를 다시 `saveDocumentContent`로 써서 최신 문서를 덮을 수 있습니다. `applyDocumentContent`에서 pending timer를 clear하고 callback을 noop으로 돌리거나, save generation/version을 두어 stale callback이 저장하지 못하게 막아야 합니다.

2. `frontend/src/graph/converter.ts:294` - 수평 leg 관통 문제는 아직 남아 있습니다. 같은 행에서 `startRight < obstacle < endLeft`인 경우 `midX`를 obstacle의 오른쪽으로 옮기면 첫 번째 horizontal segment가 노드를 지나고, 왼쪽으로 옮기면 두 번째 horizontal segment가 노드를 지납니다. 즉 `midX`만 조정하는 3-segment 경로로는 이 케이스를 피할 수 없고, 반복도 같은 `midX`에 머문 뒤 clipping 상태로 반환합니다. vertical detour를 위/아래로 추가하거나, dagre waypoint/다중 waypoint 경로를 실제로 사용해야 합니다.

3. `frontend/src/graph/converter.test.ts:239` - 새 테스트가 실제 회귀를 놓칩니다. `points`는 arrow-local 좌표라 x뿐 아니라 y도 `arrow.y`를 더해야 absolute 좌표가 됩니다. 지금은 `py`만 쓰기 때문에 실제 horizontal segment가 `y=25`에서 노드를 관통해도 테스트에서는 `y=0`으로 계산되어 `y1 > cTop` 조건이 false가 됩니다. 그래서 현재 router가 여전히 중간 노드를 통과하는데도 테스트가 통과합니다.

CI는 최신 run 기준 통과지만, 위 항목들은 PR 목표인 저장 안정성/화살표 관통 회귀 제거에 직접 걸려서 request changes로 남깁니다. 추가로 `git diff --check`에서 `frontend/src/graph/converter.ts` EOF blank line 경고도 하나 나옵니다.

**Comment by jinsh1210** [comment]
리뷰 반영했습니다. 최신 head `73f6dc4` 기준입니다.

---

**[P1] save race — `applyDocumentContent` 진입 시 pending timer clear (`App.tsx:397`)**

`applyDocumentContent` 진입부에서 `saveDebounceTimerRef`를 clear하고 `pendingSaveCallbackRef`를 noop으로 reset합니다. 이후 stale closure가 실행될 경로가 없어져서 외부 document update 200ms 이내에 예약된 local save가 최신 content를 덮는 경합이 제거됩니다. JS 단일 스레드이므로 generation counter 없이 clearTimeout 단독으로 충분합니다.

**[P1] converter.test.ts 좌표 버그 — `py + oy` (`converter.test.ts:239`)**

`oy = arrow.y`를 추가하고 `[px + ox, py + oy]`로 절대좌표를 올바르게 복원합니다. 수정 전에는 y에 offset이 없어서 `y1 > cTop=0` 조건이 false가 되어 중간 노드 관통 케이스를 놓쳤습니다. 수정 후 라우터 미수정 상태에서 해당 테스트가 fail 하는 것을 확인하고 진행했습니다.

**[P1] routeAroundNodes 수평 leg 관통 — U-shape detour fallback (`converter.ts:238`)**

`midX`를 하나씩 nudge하던 방식 대신, default midpoint + 각 obstacle의 left/right edge 목록을 후보로 두고 3-segment가 모두 clear되는 첫 번째 후보를 선택합니다. 후보가 전부 실패하면 (동일 행 케이스 포함) 5-segment U-shape(startRight→midX1→detourY→midX2→endLeft)으로 fallback해 obstacle strip 위/아래를 우회합니다. 기존 `Math.abs(leftX - midX) <= ...` 기준은 가까운 쪽이 통과 방향과 달라 1차 horizontal leg를 여전히 관통시킬 수 있었습니다.

**Review by babytazo — CHANGES_REQUESTED** [review]
3차 리뷰입니다.

최신 head `73f6dc4` 기준으로 확인했습니다. 2차에서 막았던 `applyDocumentContent` vs pending debounced save 경합은 진입부에서 timer clear + callback noop 처리로 해소됐고, `converter.test.ts`의 absolute y 복원도 `py + oy`로 고쳐졌습니다. 최신 GitHub Actions도 Node/unit, Node/integration, extension-e2e 모두 성공했고, 로컬에서도 PR head로 `frontend`의 `converter.test.ts` + `canvasStorage.test.ts`를 돌려 26개 테스트 통과를 확인했습니다.

다만 아직 한 가지 라우팅 edge case가 남아 request changes로 둡니다.

`frontend/src/graph/converter.ts:321-331` - U-shape fallback이 만든 경로를 다시 obstacle에 대해 검증하지 않습니다. 현재 fallback은 `midX1 = startRight + gap`, `midX2 = endLeft - gap`으로 고정한 뒤 바로 반환하는데, obstacle이 source 바로 오른쪽에 붙어 있는 preserved/manual position 케이스에서는 첫 horizontal leg와 첫 vertical riser가 obstacle 내부를 지나갑니다. 예를 들어 source가 `x=0,w=200,y=0,h=60`, target이 `x=600`, obstacle이 `x=205,y=0,w=200,h=60`이면 fallback 경로는 `(200,30)->(212,30)->(212,-12)...`가 되고, `(200,30)->(212,30)` 및 `x=212` vertical segment가 obstacle 범위 `x=205..405,y=0..60`을 관통합니다.

이 fallback도 반환 전에 전체 polyline segment를 검증하고, 실패 시 obstacle strip 바깥의 안전한 x 좌표를 선택하거나 다중 waypoint 후보를 재시도해야 합니다. 기존 positions를 보존하는 구조라 사용자가 auto node를 이동시킨 저장 장면에서도 재현 가능한 회귀라서, 이 케이스를 테스트로 추가하는 편이 좋겠습니다.

추가로 `git diff --check origin/develop..origin/pr-100`에서 `frontend/src/graph/converter.ts:454: new blank line at EOF`가 아직 남아 있으니 함께 정리 부탁드립니다.

**Comment by jinsh1210** [comment]
리뷰 반영했습니다. 최신 head `ec6d536` 기준입니다.

---

**[P1] U-shape fallback midX1/midX2 고정값 관통 — candidate probe로 교체 (`converter.ts:319`)**

`midX1 = startRight + gap`, `midX2 = endLeft - gap` 고정 대신 후보 목록을 탐색합니다. `midX1Candidates`에 `startRight + gap`, **`startRight` (zero-length first h-seg)**, 각 obstacle의 right edge + gap을 추가했습니다. `startRight` 자체를 후보로 넣으면 source 바로 오른쪽에 obstacle이 붙어 있는 케이스(첫 horizontal leg가 피할 x가 없는 경우)에서 첫 segment 길이를 0으로 만들어 관통 없이 detourY로 수직 탈출합니다. 각 후보는 기존 `polylineClear`로 전체 segment를 검증한 뒤 반환합니다.

**[nit] EOF blank line (`converter.ts:470`)**

trailing blank line 1개 제거했습니다.

**회귀 테스트 추가 (`converter.test.ts:268`)**

리뷰어 예제 케이스(source x=0,w=200 / target x=600 / obstacle x=205,y=0,w=200,h=60)를 테스트로 추가했습니다. 라우터 수정 전 fail, 수정 후 pass 확인하고 진행했습니다.

**Review by babytazo — APPROVED** [review]
4차 리뷰입니다.

최신 head `ec6d536` 기준으로 다시 확인했습니다. 3차에서 남겼던 U-shape fallback의 source-adjacent obstacle 관통 케이스는 `polylineClear` 기반 후보 검증으로 정리됐고, 해당 회귀 테스트도 추가된 것을 확인했습니다. `git diff --check`의 EOF blank line 경고도 더 이상 나오지 않습니다.

추가 확인:
- GitHub Actions: 최신 커밋 기준 Node/unit, Node/integration, extension-e2e 모두 성공
- 로컬: PR head에서 `npm test --workspace=frontend -- --runInBand` 통과 (`79 passed`)
- 로컬: PR head에서 `npm run build --workspace=frontend` 통과

이전 리뷰에서 막았던 저장 debounce race, pan/zoom false-skip, saved z-order, sticky connector repair, arrow routing 회귀까지 모두 충분히 수습됐다고 봅니다. Approve합니다.
